### PR TITLE
AP_Math: always check math indexes in SITL

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by tridge** [comment]
that means our normal code paths are never checked ... not sure I like that

**Comment by tridge** [comment]
I think we should have a command line flag to sim_vehicle.py which enables this, and enable it on some CI tests, but not all

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -288,6 +288,9 @@ def configure_env(self, cfg, env):
                 if fnmatch.fnmatch(f, "font*bin"):
                     env.ROMFS_FILES += [(f,'libraries/AP_OSD/fonts/'+f)]
 
+        if cfg.options.enable_math_check_indexes:
```
Comment (OXINARF) [inline Tools/ardupilotwaf/boards.py]: Why not put this for all boards? It's disabled by default so...

**Comment by peterbarker** [comment]
I've added a patch to actually use this in CI now - for the Rover test only.

File: libraries/APM_Control/AP_PitchController.cpp [hunk 2]
```diff
@@ -211,7 +211,7 @@ int32_t AP_PitchController::_get_rate_out(float desired_rate, float scaler, bool
       beyond the configured roll limit, reducing to zero at 90
       degrees
     */
-    float roll_wrapped = fabsf(_ahrs.roll_sensor);
+    float roll_wrapped = labs(_ahrs.roll_sensor);
```
Comment (khancyr) [inline libraries/APM_Control/AP_PitchController.cpp]: change it int32_t maybe ?

**Comment by peterbarker** [comment]
Must be checked on stm32; use disable interrupts, do a perf reset on the nsh console (with compass and GPS - fusing everything), then do the same thing with this enabled.

This can be merged if the results are similar.

**Comment by tridge** [comment]
needs performance testing on stm32 to ensure we get equivalent EKF performance

**Comment by OXINARF** [comment]
@peterbarker Please look at my latest review, it's not fixed, as far as I can tell.

**Comment by peterbarker** [comment]
@OXINARF - you want the option moved up before the linux options?

**Comment by peterbarker** [comment]
@OXINARF Never mind.  I was confused by the "scripting too" thing.  FWIU @WickedShell only wanted scripting on for Linux for the time being.  Will move the option appropriately.

**Comment by peterbarker** [comment]
master:

```
EK2_FuseOptFlow: 23023 events, 0 overruns, 60579us elapsed, 2us avg, min 1us max 259us 12.174us rms
EK2_TerrainOffset: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseSideslip: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseAirspeed: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseMagnetometer: 23024 events, 0 overruns, 1082790us elapsed, 47us avg, min 2us max 636us 86.895us rms
EK2_FuseVelPosNED: 5984 events, 0 overruns, 2268298us elapsed, 379us avg, min 100us max 1056us 234.713us rms
EK2_CovariancePrediction: 23024 events, 0 overruns, 4345716us elapsed, 188us avg, min 111us max 557us 83.089us rms
EK2_UpdateFilter: 23024 events, 0 overruns, 11231469us elapsed, 487us avg, min 208us max 1663us 251.986us rms
```


master + this patch:

```
EK2_FuseOptFlow: 36035 events, 0 overruns, 88426us elapsed, 2us avg, min 1us max 289us 11.765us rms
EK2_TerrainOffset: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseSideslip: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseAirspeed: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseMagnetometer: 36046 events, 0 overruns, 1765054us elapsed, 48us avg, min 2us max 576us 93.376us rms
EK2_FuseVelPosNED: 9239 events, 0 overruns, 3338808us elapsed, 361us avg, min 100us max 1037us 235.500us rms
EK2_CovariancePrediction: 36046 events, 0 overruns, 6431774us elapsed, 178us avg, min 111us max 655us 103.403us rms
EK2_UpdateFilter: 36046 events, 0 overruns, 17662747us elapsed, 490us avg, min 208us max 1670us 243.815us rms
EK2_FuseVelPosNED: 9242 events, 0 overruns, 3537655us elapsed, 382us avg, min 94us max 1067us 233.359us rms
EK2_CovariancePrediction: 36062 events, 0 overruns, 7515780us elapsed, 208us avg, min 107us max 692us 85.840us rms
EK2_UpdateFilter: 36062 events, 0 overs elapsed, 19653us avg, min 2982us max 26993us 1092.811us rms
```


Deltas appear to be well within the relevant rms's.

**Comment by OXINARF** [comment]
> FWIU @WickedShell only wanted scripting on for Linux for the time being.

@peterbarker The script option is used for all boards (https://github.com/ArduPilot/ardupilot/blob/master/Tools/ardupilotwaf/boards.py#L94) so it should also be displayed in the help text in the correct place. Anyway, I was only requesting your new option to be in the correct place, not your job to fix script 😉

> (editing to correct corrupt new-output)

Still seems strange, several ones are repeated.

**Comment by peterbarker** [comment]
This time with interrupts disabled:

{code}
master / no interrupts:

EK2_FuseOptFlow: 13241 events, 0 overruns, 21995us elapsed, 1us avg, min 1us max 2us 0.473us rms
EK2_TerrainOffset: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseSideslip: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseAirspeed: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseMagnetometer: 13252 events, 0 overruns, 382594us elapsed, 28us avg, min 2us max 355us 54.559us rms
EK2_FuseVelPosNED: 2651 events, 0 overruns, 869249us elapsed, 327us avg, min 322us max 329us 0.942us rms
EK2_CovariancePrediction: 13252 events, 0 overruns, 1642749us elapsed, 123us avg, min 123us max 124us 0.191us rms
EK2_UpdateFilter: 13252 events, 0 overruns, 4276031us elapsed, 322us avg, min 22000us rms


math-indexes / no-interrupts:


EK2_FuseOptFlow: 14568 events, 0 overruns, 23566us elapsed, 1us avg, min 1us max 2us 0.486us rms
EK2_TerrainOffset: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseSideslip: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseAirspeed: 0 events, 0 overruns, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
EK2_FuseMagnetometer: 14569 events, 0 overruns, 510865us elapsed, 35us avg, min 2us max 358us 59.849us rms
EK2_FuseVelPosNED: 2914 events, 0 overruns, 952906us elapsed, 327us avg, min 322us max 329us 1.113us rms
EK2_CovariancePrediction: 14569 events, 0 overruns, 1811609us elapsed, 124us avg, min 124us max 125us 0.476us rms
EK2_UpdateFilter: 14569 events, 0 overruns, 4800248us elapsed, 329us avg, min 220us max 903us 126.026us rms
{code}

**Comment by peterbarker** [comment]
On Tue, 16 Oct 2018, Francisco Ferreira wrote:

> Still seems strange, several ones are repeated.

Yes!  I'm assuming some sort of race condition in the perf stuff ATM.

**Comment by tridge** [comment]
I think this can be merged when you are ready @peterbarker

**Comment by peterbarker** [comment]
On Wed, 17 Oct 2018, Francisco Ferreira wrote:

>       argument moved out from underneath Linux
> 
> @peterbarker Did you forget to force-push? It is still in the same place.

Force-pushed now.

**Pull request merged**